nui/src/main/java/org/terasology/nui/widgets/types/builtin/object/ObjectLayoutBuilder.java

```diff
@@ -0,0 +1,211 @@
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+package org.terasology.nui.widgets.types.builtin.object;
+
+import com.google.common.base.Defaults;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.nui.UIWidget;
+import org.terasology.nui.WidgetUtil;
+import org.terasology.nui.databinding.Binding;
+import org.terasology.nui.databinding.DefaultBinding;
+import org.terasology.nui.databinding.NotifyingBinding;
+import org.terasology.nui.itemRendering.StringTextRenderer;
+import org.terasology.nui.layouts.ColumnLayout;
+import org.terasology.nui.widgets.UIBox;
+import org.terasology.nui.widgets.UIButton;
+import org.terasology.nui.widgets.UIDropdownScrollable;
+import org.terasology.nui.widgets.UILabel;
+import org.terasology.nui.widgets.types.TypeWidgetLibrary;
+import org.terasology.nui.widgets.types.builtin.util.ExpandableLayoutBuilder;
+import org.terasology.nui.widgets.types.builtin.util.FieldsWidgetBuilder;
+import org.terasology.reflection.ReflectionUtil;
+import org.terasology.reflection.TypeInfo;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Parameter;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.terasology.nui.widgets.types.TypeWidgetFactory.LABEL_WIDGET_ID;
+
+public class ObjectLayoutBuilder<T> extends ExpandableLayoutBuilder<T> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ObjectLayoutBuilder.class);
+    private static final String NULL_LABEL = "Object is null.";
+    private static final String MODIFY_LABEL = "Modify Object";
+
+    private final TypeInfo<T> type;
+    private final TypeWidgetLibrary library;
+    private final FieldsWidgetBuilder<T> fieldsWidgetBuilder;
+
+    public ObjectLayoutBuilder(TypeInfo<T> type, TypeWidgetLibrary library) {
+        this.type = type;
+        this.library = library;
+        fieldsWidgetBuilder = new FieldsWidgetBuilder<>(type, library);
+    }
+
+    @Override
+    protected void populate(Binding<T> binding, ColumnLayout layout, ColumnLayout mainLayout) {
+        layout.removeAllWidgets();
+
+        UILabel nameWidget = mainLayout.find(LABEL_WIDGET_ID, UILabel.class);
+        assert nameWidget != null;
+
+        if (binding.get() == null) {
+            populateNullLayout(binding, layout, nameWidget);
+        } else {
+            buildEditorLayout(binding, layout, nameWidget);
+        }
+    }
+
+    private void buildEditorLayout(Binding<T> binding, ColumnLayout layout, UILabel nameWidget) {
+        // TODO: Translate
+        if (NULL_LABEL.equals(nameWidget.getText())) {
+            nameWidget.setText(MODIFY_LABEL);
```

Comment:
Since this is not actually being used anywhere right now, I think it will be better to address this issue in a later, more specialized PR than add to this already monolithic PR. IIRC there are a few other strings in UI code introduced by this PR that also require internationalization -- it might be better to just fix them all in one go later.

Reply:
I brought this up during office hours today, and it seems others agree: at this time, with the current level of API churn, better to merge this as it is and revisit details like i18n after we no longer need to maintain two different major versions of TeraNUI.